system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,162 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitDetails, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitSubject, g_object_unref)
 
+typedef struct
+{
+  FlatpakSystemHelper *object;
+  GDBusMethodInvocation *invocation;
+  GCancellable *cancellable;
+
+  guint watch_id;
+  uid_t uid;
+
+  gchar *root_dir_name;
+  gchar *user_dir_name;
+
+  GUnixMountMonitor *mount_monitor;
+  gulong mount_monitor_id;
+  guint unsuccessful_mount_timeout_id;
+  GSubprocess *bindfs;
+} FlatpakSystemHelperOngoingPull;
+
+static void
+system_helper_ongoing_pull_free (FlatpakSystemHelperOngoingPull *pull)
+{
+  g_autoptr(GFile) root_dir = g_file_new_for_path (pull->root_dir_name);
+  g_autoptr(GFile) user_dir = g_file_new_for_path (pull->user_dir_name);
+
+  g_cancellable_cancel (pull->cancellable);
```

Comment:
>Is there a chance it could be freed in unsuccessful_mount_timeout and then again in handle_deploy?

No, it can't happen. `unsuccessful_mount_timeout` will return the error and the pull will fallback to a `/var/tmp/..` location; and in the `handle_deploy`, it gets the fallback path and fails `ongoing_pull != NULL` condition. 

>I think there's a race because system_helper_ongoing_pull_free frees the members of the struct while the pull is still in the ongoing_pulls array, where another thread might find it and use or free it.

Still having trouble what scenario might run into this. Maybe something like this is better?
```
system_helper_ongoing_pull_free ( pull )
{
  gboolean removed = FALSE;
  G_LOCK (ongoing_pulls);
  removed = g_ptr_remove_array (ongoing_pulls, pull);
  G_UNLOCK (ongoing_pulls);

  if (removed)
  /* Free other members */
}
```

Reply:
That check is too late: if two threads are operating on the same pointer, and one calls `free` before the other, then the other will be accessing freed memory.

To see a case where two threads may be operating on the same pointer, consider what would happen if a client calls `Deploy` and `CancelPull` for the same path at the same time.

 I think the following changes make sense:

* Split the bits of `_free` which are about tearing down state rather than freeing stuff into a separate `_teardown` or `_stop` function
* Add a refcount to the struct
* Take refs in the cases where the struct is passed as user_data to an async function
* Explicitly call `_teardown` on error paths, and also call it from `_unref` when we reach `0` refs and free the contents
* Make `get_ongoing_pull_by_dir` remove the pull from the array before returning it

Reply:
> Make get_ongoing_pull_by_dir remove the pull from the array before returning it

Won't this break the criteria of idle exit ? idle timeout might see `ongoing_pulls` as none, and exits the system helper?

Reply:
> Split the bits of _free which are about tearing down state rather than freeing stuff into a separate _teardown or _stop function
> Add a refcount to the struct
> Take refs in the cases where the struct is passed as user_data to an async function
> Explicitly call _teardown on error paths, and also call it from _unref when we reach 0 refs and free the contents

Addressed these.

> Make get_ongoing_pull_by_dir remove the pull from the array before returning it

Didn't address this (yet). I constantly find myself out of brain power to figure out any other approach here. I will probably need to take a fresh look after sometime.